# Mengatur *Environment*

In [ ]:
# Menginstall dependencies
!pip install pypdf -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 249.5/249.5 kB 6.3 MB/s eta 0:00:00


In [ ]:
# Mengimpor modules
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

import os, pathlib, string, re, pickle, json

from pypdf import PdfReader

# UU 13/2003 tentang Ketenagakerjaan

In [ ]:
# Menginisialisasi list yang akan mengandung seluruh potongan peraturan
UU_13_2003 = []

## Text Extraction

In [ ]:
# Memilih sumber data (dokumen pdf)
file_name = os.path.basename('/content/UU Nomor 13 Tahun 2003.pdf')
judul_UU_13_2003 = pathlib.Path(file_name).stem
judul_UU_13_2003

'UU Nomor 13 Tahun 2003'

In [ ]:
# Dimodifikasi dari https://pypdf2.readthedocs.io/en/latest/user/extract-text.html
# Mengekstrak teks dari suatu dokumen
extracted_pdf_with_pypdf = ""
reader = PdfReader(file_name)

# Helper function untuk mengabaikan header dan footer
def visitor_body(text, cm, tm, fontDict, fontSize):
    y = tm[5]
    if y > 85 and y < 815: # dari bawah, lebih dari 85 dan tidak sampai 815
        parsed_parts.append(text)

for i in range(len(reader.pages)):
  parsed_parts = []
  reader.pages[i].extract_text(visitor_text=visitor_body)
  whole_page = "".join(parsed_parts)
  extracted_pdf_with_pypdf += whole_page

print(extracted_pdf_with_pypdf)


UNDANG -UNDANG REPUBLIK INDONESIA
NOMOR 13 TAHUN 2003
TENTANG
KETENAGAKERJAAN
DENGAN RAHMAT TUHAN YANG MAHA ESA
PRESIDEN REPUBLIK INDONESIA,
Menimbang :a. bahwa pembangunan nasional dilaksanakan dalam rangka
pembangunan manusia Indonesia seutuhnya dan pembangunan
masyarakat Indonesia seluruhnya untuk mewujudkan masyarakat yang
sejahtera, adil, makmur, yang merata, baik materiil maupun spiritual
berdasarkan P ancasila dan Undang Undang Dasar Negara Republik
Indonesia Tahun 1945;
b. bahwa dalam pelaksanaan pembangunan nasional, tenaga kerja
mempunyai peranan dan kedudukan yang sangat penting sebagai
pelaku dan tujuan pembangunan;
c. bahwa sesuai dengan peranan d an kedudukan tenaga kerja, diperlukan
pembangunan ketenagakerjaan untuk meningkatkan kualitas tenaga
kerja dan peransertanya dalam pembangunan serta peningkatan
perlindungan tenaga kerja dan keluarganya sesuai dengan harkat dan
martabat kemanusiaan;
d. bahwa perlindungan terhadap tenaga kerja dimaksudkan untuk
menjamin hak hak d

## Text Preprocessing

In [ ]:
extracted_pdf_with_pypdf

'\nUNDANG -UNDANG REPUBLIK INDONESIA\nNOMOR 13 TAHUN 2003\nTENTANG\nKETENAGAKERJAAN\nDENGAN RAHMAT TUHAN YANG MAHA ESA\nPRESIDEN REPUBLIK INDONESIA,\nMenimbang :a. bahwa pembangunan nasional dilaksanakan dalam rangka\npembangunan manusia Indonesia seutuhnya dan pembangunan\nmasyarakat Indonesia seluruhnya untuk mewujudkan masyarakat yang\nsejahtera, adil, makmur, yang merata, baik materiil maupun spiritual\nberdasarkan P ancasila dan Undang Undang Dasar Negara Republik\nIndonesia Tahun 1945;\nb. bahwa dalam pelaksanaan pembangunan nasional, tenaga kerja\nmempunyai peranan dan kedudukan yang sangat penting sebagai\npelaku dan tujuan pembangunan;\nc. bahwa sesuai dengan peranan d an kedudukan tenaga kerja, diperlukan\npembangunan ketenagakerjaan untuk meningkatkan kualitas tenaga\nkerja dan peransertanya dalam pembangunan serta peningkatan\nperlindungan tenaga kerja dan keluarganya sesuai dengan harkat dan\nmartabat kemanusiaan;\nd. bahwa perlindungan terhadap tenaga kerja dimaksudkan un

In [ ]:
# Mempartisi bagian penjelasan
partioned_string_1, partioned_string_2, partioned_string_3  = extracted_pdf_with_pypdf.partition("\nPENJELASAN\nATAS")
bagian_penjelasan = partioned_string_2 + partioned_string_3
bagian_penjelasan

'\nPENJELASAN\nATAS\nUNDANG -UNDANG REPUBLIK INDONESIA\nNOMOR 13 TAHUN 2003\nTENTANG\nKETENAGAKERJAAN\nI.UMUM\nPembangunan ketenagakerjaan sebagai bagian integral dari pembangunan\nnasional berdasarkan Pancasila dan Undang -undang Dasar Negara Republik Indonesia\nTahun 1945, dilak sanakan dalam rangka pembangunan manusia Indonesia seutuhnya\ndan pembangunan masyarakat Indonesia seluruhnya untuk meningkatkan harkat,\nmartabat, dan harga diri tenaga kerja serta mewujudkan masyarakat sejahtera, adil,\nmakmur, dan merata, baik materiil mau pun spiritual.\nPembangunan ketenagakerjaan harus diatur sedemikian rupa sehingga\nterpenuhi hak -hak dan perlindungan yang mendasar bagi tenaga kerja dan\npekerja/buruh serta pada saat yang bersamaan dapat mewujudkan kondisi yang\nkondusif bagi pengembangan dunia usaha.\nPembangunan ketenagakerjaan mempunyai banyak dimensi dan keterkaitan.\nKeterkaitan itu tidak hanya dengan kepentingan tenaga kerja selama, sebelum dan\nsesudah masa kerja tetapi juga ke

In [ ]:
# Mempartisi bagian penutup
partioned_string_1, partioned_string_2, partioned_string_3 = partioned_string_1.partition("Disahkan di")
bagian_penutup = partioned_string_2 + partioned_string_3
bagian_penutup

'Disahkan di Jakarta\npada tanggal 25 Maret 2003\nPRESIDEN REPUBLIK INDONESIA,\nttd\nMEGAWATI SOEKARNOPUTRI\nLEMBARA N NEGARA REPUBLIK INDONESIA TAHUN 2003 NOMOR 39'

In [ ]:
# Mempartisi bagian batang tubuh
partioned_string_1, partioned_string_2, partioned_string_3 = partioned_string_1.partition("BAB I")
bagian_batang_tubuh = partioned_string_2 + partioned_string_3
bagian_batang_tubuh

'BAB I\nKETENTUAN UMUM\nPasal 1\nDalam undang undang ini yang dimaksud dengan :\n1. Ketenagakerjaan adalah segala hal yang berhubungan dengan tenaga kerja\npada waktu sebelum, selama, dan sesudah masa kerja.\n2. Tenaga kerja adalah setiap orang yang mampu melakukan pekerjaan guna\nmenghasilkan barang dan/atau jasa baik untuk memenuhi kebutuh an\nsendiri maupun untuk masyarakat.\n3. Pekerja/buruh adalah setiap orang yang bekerja dengan menerima upah\natau imbalan dalam bentuk lain.\n4. Pemberi kerja adalah orang perseorangan, pengusaha, bada n hukum, atau\nbadan -badan lainnya yang mempekerjakan tenaga kerja dengan membayar\nupah atau imbalan dalam bentuk lain.\n5. Pengusaha adalah :\na.orang perseorangan, persekutuan, atau badan hukum yang menjalankan\nsuatu perusahaan milik sendiri;\n-3-\nb.orang pers eorangan, persekutuan, atau badan hukum yang secara\nberdiri sendiri menjalankan perusahaan bukan miliknya;\nc.orang perseorangan, persekutuan, atau badan hukum yang berada di\nIndonesia

### Bagian batang tubuh

In [ ]:
# Fungsi untuk menghilangkan bagian yang tidak diinginkan setelah parsing
def text_preprocessing_1(input):
  temp = re.sub('\\nBAB.*?\s\.\.\.\\n-\d+-', '', input)
  temp = re.sub('\\nBagian.*?\s\.\.\.\\n-\d+-', '', temp)
  temp = re.sub('\\nParagraf.*?\s\.\.\.\\n-\d+-', '', temp)
  temp = re.sub('\\nPasal.*?\s\.\.\.\\n-\d+-', '', temp)
  temp = re.sub('\\n\(\d+\).*?…\\n-\d+-', '', temp)
  temp = re.sub('\\n\d+.*?\.\.\.\\n-\d+-', '', temp)
  temp = re.sub('\\n[a-z].*?…\\n-\d+-', '', temp)
  output = re.sub('\\n-\d+-', '', temp)
  return output

In [ ]:
# Mengurai keseluruhan teks berdasarkan nomor pasal yang ada (parsing)
pattern = "\\nPasal\s\d+\\n"
list_pasal = []
list_pasal =  [result for result in re.split(pattern, text_preprocessing_1(bagian_batang_tubuh))]

# Memodifikasi teks (menambahkan identifier tertentu)
# Harapannya suatu LLM dapat diinstruksikan untuk mengenali isi dari tiap nomor pasal berdasarkan identifier tersebut
for i in range(len(list_pasal)):
  list_pasal[i] = "[Berikut adalah isi Pasal " + str(i) + " pada " + judul_UU_13_2003 + "] " + list_pasal[i]

len(list_pasal)

194

In [ ]:
# Memeriksa isi dari list pasal yang telah dibuat
print(list_pasal[0])

[Berikut adalah isi Pasal 0 pada UU Nomor 13 Tahun 2003] BAB I
KETENTUAN UMUM


In [ ]:
# Memeriksa isi dari list pasal yang telah dibuat
print(list_pasal[1])

[Berikut adalah isi Pasal 1 pada UU Nomor 13 Tahun 2003] Dalam undang undang ini yang dimaksud dengan :
1. Ketenagakerjaan adalah segala hal yang berhubungan dengan tenaga kerja
pada waktu sebelum, selama, dan sesudah masa kerja.
2. Tenaga kerja adalah setiap orang yang mampu melakukan pekerjaan guna
menghasilkan barang dan/atau jasa baik untuk memenuhi kebutuh an
sendiri maupun untuk masyarakat.
3. Pekerja/buruh adalah setiap orang yang bekerja dengan menerima upah
atau imbalan dalam bentuk lain.
4. Pemberi kerja adalah orang perseorangan, pengusaha, bada n hukum, atau
badan -badan lainnya yang mempekerjakan tenaga kerja dengan membayar
upah atau imbalan dalam bentuk lain.
5. Pengusaha adalah :
a.orang perseorangan, persekutuan, atau badan hukum yang menjalankan
suatu perusahaan milik sendiri;
b.orang pers eorangan, persekutuan, atau badan hukum yang secara
berdiri sendiri menjalankan perusahaan bukan miliknya;
c.orang perseorangan, persekutuan, atau badan hukum yang berada di
Indones

In [ ]:
# Memeriksa isi dari list pasal yang telah dibuat
print(list_pasal[2])

[Berikut adalah isi Pasal 2 pada UU Nomor 13 Tahun 2003] Pembangunan ketenagakerjaan berlandaskan Pancasila dan Undang Undang
Dasar Negara Republik Indonesia Tahun 1945.


In [ ]:
# Fungsi untuk 'membersihkan' teks
def text_preprocessing_2(input):
  text = re.sub(r'\n', ' ', input) # mengganti jarak antar baris menjadi sebuah spasi
  text = re.sub(r'BAB.*', '', text) # menghilangkan bagian "BAB ..." yang terdeteksi saat parsing
  text = re.sub(r'Bagian.*', '', text) # menghilangkan bagian "Bagian ..." yang terdeteksi saat parsing
  text = re.sub(r'Paragraf.*', '', text) # menghilangkan bagian "Paragraf ..." yang terdeteksi saat parsing

  # menghilangkan jarak antar baris yang berlebih
  text = re.sub(' +', ' ', text.replace('\n', ' '))
  output = text.strip()
  return output

In [ ]:
list_pasal_cleaned = []
for i in range(len(list_pasal)):
  list_pasal_cleaned.append(text_preprocessing_2(list_pasal[i]))

len(list_pasal_cleaned)

194

***parse* ulang Pasal 1 (ketentuan umum) karena mengandung terlalu banyak konteks**

In [ ]:
# ketentuan umum
partioned_string_1, partioned_string_2, partioned_string_3  = bagian_batang_tubuh.partition("BAB II")
bagian_ketentuan_umum = partioned_string_1
bagian_ketentuan_umum

'BAB I\nKETENTUAN UMUM\nPasal 1\nDalam undang undang ini yang dimaksud dengan :\n1. Ketenagakerjaan adalah segala hal yang berhubungan dengan tenaga kerja\npada waktu sebelum, selama, dan sesudah masa kerja.\n2. Tenaga kerja adalah setiap orang yang mampu melakukan pekerjaan guna\nmenghasilkan barang dan/atau jasa baik untuk memenuhi kebutuh an\nsendiri maupun untuk masyarakat.\n3. Pekerja/buruh adalah setiap orang yang bekerja dengan menerima upah\natau imbalan dalam bentuk lain.\n4. Pemberi kerja adalah orang perseorangan, pengusaha, bada n hukum, atau\nbadan -badan lainnya yang mempekerjakan tenaga kerja dengan membayar\nupah atau imbalan dalam bentuk lain.\n5. Pengusaha adalah :\na.orang perseorangan, persekutuan, atau badan hukum yang menjalankan\nsuatu perusahaan milik sendiri;\n-3-\nb.orang pers eorangan, persekutuan, atau badan hukum yang secara\nberdiri sendiri menjalankan perusahaan bukan miliknya;\nc.orang perseorangan, persekutuan, atau badan hukum yang berada di\nIndonesia

In [ ]:
list_bagian_ketentuan_umum = re.split('\\n\d+\.\s', bagian_ketentuan_umum)
len(list_bagian_ketentuan_umum)

37

> Pasal 1 seharusnya terdiri dari 33 angka

In [ ]:
list_bagian_ketentuan_umum

['BAB I\nKETENTUAN UMUM\nPasal 1\nDalam undang undang ini yang dimaksud dengan :',
 'Ketenagakerjaan adalah segala hal yang berhubungan dengan tenaga kerja\npada waktu sebelum, selama, dan sesudah masa kerja.',
 'Tenaga kerja adalah setiap orang yang mampu melakukan pekerjaan guna\nmenghasilkan barang dan/atau jasa baik untuk memenuhi kebutuh an\nsendiri maupun untuk masyarakat.',
 'Pekerja/buruh adalah setiap orang yang bekerja dengan menerima upah\natau imbalan dalam bentuk lain.',
 'Pemberi kerja adalah orang perseorangan, pengusaha, bada n hukum, atau\nbadan -badan lainnya yang mempekerjakan tenaga kerja dengan membayar\nupah atau imbalan dalam bentuk lain.',
 'Pengusaha adalah :\na.orang perseorangan, persekutuan, atau badan hukum yang menjalankan\nsuatu perusahaan milik sendiri;\n-3-\nb.orang pers eorangan, persekutuan, atau badan hukum yang secara\nberdiri sendiri menjalankan perusahaan bukan miliknya;\nc.orang perseorangan, persekutuan, atau badan hukum yang berada di\nIndonesi

> Entri pada indeks 0, 12, 21, dan 33 perlu dihilangkan dari list

In [ ]:
a, b, c, d = list_bagian_ketentuan_umum.pop(33), list_bagian_ketentuan_umum.pop(21), list_bagian_ketentuan_umum.pop(12), list_bagian_ketentuan_umum.pop(0)
print("0th. {},\n12th. {},\n21th. {},\n33th. {}".format(a, b, c, d))

0th. Kesejahteraan ...
-6-,
12th. Peraturan …
-5-,
21th. Pelayanan …
-4-,
33th. BAB I
KETENTUAN UMUM
Pasal 1
Dalam undang undang ini yang dimaksud dengan :


In [ ]:
list_bagian_ketentuan_umum

['Ketenagakerjaan adalah segala hal yang berhubungan dengan tenaga kerja\npada waktu sebelum, selama, dan sesudah masa kerja.',
 'Tenaga kerja adalah setiap orang yang mampu melakukan pekerjaan guna\nmenghasilkan barang dan/atau jasa baik untuk memenuhi kebutuh an\nsendiri maupun untuk masyarakat.',
 'Pekerja/buruh adalah setiap orang yang bekerja dengan menerima upah\natau imbalan dalam bentuk lain.',
 'Pemberi kerja adalah orang perseorangan, pengusaha, bada n hukum, atau\nbadan -badan lainnya yang mempekerjakan tenaga kerja dengan membayar\nupah atau imbalan dalam bentuk lain.',
 'Pengusaha adalah :\na.orang perseorangan, persekutuan, atau badan hukum yang menjalankan\nsuatu perusahaan milik sendiri;\n-3-\nb.orang pers eorangan, persekutuan, atau badan hukum yang secara\nberdiri sendiri menjalankan perusahaan bukan miliknya;\nc.orang perseorangan, persekutuan, atau badan hukum yang berada di\nIndonesia mewakili perusahaan sebagaimana dimaksud dalam huruf a\ndan b yang berke dudukan 

In [ ]:
# Fungsi untuk 'membersihkan' teks
def text_preprocessing_3(input):
  # menghilangkan jarak antar baris yang berlebih
  text = re.sub(' +', ' ', input.replace('\n', ' '))
  output = text.strip()
  return output

In [ ]:
list_bagian_ketentuan_umum_cleaned = [text_preprocessing_3(each_item) for each_item in list_bagian_ketentuan_umum]

# modify the text so that (hopefully) the LLM can be instructed to easily identify the article number
for i in range(len(list_bagian_ketentuan_umum_cleaned)):
  list_bagian_ketentuan_umum_cleaned[i] = "[Berikut adalah isi Pasal 1 angka " + str(i+1) + " pada " + judul_UU_13_2003 + "] " + list_bagian_ketentuan_umum_cleaned[i]

len(list_bagian_ketentuan_umum_cleaned)

33

In [ ]:
# Memuat isi pasal 1 sampai dengan 193 ke dalam list yang telah diinisialisasi di awal
# Pasal 1 perlu dikecualikan sebab akan dipecah lagi isinya --> maka mulai dari indeks 2
UU_13_2003 = list_bagian_ketentuan_umum_cleaned + list_pasal_cleaned[2:]
UU_13_2003

['[Berikut adalah isi Pasal 1 angka 1 pada UU Nomor 13 Tahun 2003] Ketenagakerjaan adalah segala hal yang berhubungan dengan tenaga kerja pada waktu sebelum, selama, dan sesudah masa kerja.',
 '[Berikut adalah isi Pasal 1 angka 2 pada UU Nomor 13 Tahun 2003] Tenaga kerja adalah setiap orang yang mampu melakukan pekerjaan guna menghasilkan barang dan/atau jasa baik untuk memenuhi kebutuh an sendiri maupun untuk masyarakat.',
 '[Berikut adalah isi Pasal 1 angka 3 pada UU Nomor 13 Tahun 2003] Pekerja/buruh adalah setiap orang yang bekerja dengan menerima upah atau imbalan dalam bentuk lain.',
 '[Berikut adalah isi Pasal 1 angka 4 pada UU Nomor 13 Tahun 2003] Pemberi kerja adalah orang perseorangan, pengusaha, bada n hukum, atau badan -badan lainnya yang mempekerjakan tenaga kerja dengan membayar upah atau imbalan dalam bentuk lain.',
 '[Berikut adalah isi Pasal 1 angka 5 pada UU Nomor 13 Tahun 2003] Pengusaha adalah : a.orang perseorangan, persekutuan, atau badan hukum yang menjalankan su

### Bagian penjelasan

In [ ]:
# Mengurai keseluruhan teks berdasarkan nomor pasal yang ada (parsing)
pattern = "\\nPasal\s\d+\\n"
list_penjelasan_pasal =  [result for result in re.split(pattern, text_preprocessing_1(bagian_penjelasan))]
len(list_penjelasan_pasal)

193

> Terjadi kesalahan pada parsing, sebab seharusnya panjang list tersebut adalah 194

Setelah saya cek secara manual, ternyata kesalahan parsing-nya adalah `\\nPasal 7 8`. Terdapat spasi di antara 7 dan 8 sehingga tidak termasuk ke dalam seperator pattern.

In [ ]:
# Fungsi untuk menghilangkan bagian yang tidak diinginkan saat parsing
def text_preprocessing_4(input):
  temp = re.sub('\\nBAB.*?\s\.\.\.\\n-\d+-', '', input)
  temp = re.sub('\\nBagian.*?\s\.\.\.\\n-\d+-', '', temp)
  temp = re.sub('\\nParagraf.*?\s\.\.\.\\n-\d+-', '', temp)
  temp = re.sub('\\nPasal.*?\s\.\.\.\\n-\d+-', '', temp)
  temp = re.sub('\\n\(\d+\).*?…\\n-\d+-', '', temp)
  temp = re.sub('\\n\d+.*?\.\.\.\\n-\d+-', '', temp)
  temp = re.sub('\\n[a-z].*?…\\n-\d+-', '', temp)
  temp = re.sub('\\n-\d+-', '', temp)
  output = re.sub('\\nPasal 7 8', '\\nPasal 78\\n', temp)
  return output

In [ ]:
# Mengurai keseluruhan teks berdasarkan nomor pasal yang ada (parsing)
pattern = "\\nPasal\s\d+\\n"
list_penjelasan_pasal =  [result for result in re.split(pattern, text_preprocessing_4(bagian_penjelasan))]
len(list_penjelasan_pasal)

194

In [ ]:
list_penjelasan_pasal_cleaned = [text_preprocessing_3(each_item) for each_item in list_penjelasan_pasal]

# Memodifikasi teks (menambahkan identifier tertentu)
# Harapannya suatu LLM dapat diinstruksikan untuk mengenali isi dari tiap nomor pasal berdasarkan identifier tersebut
for i in range(len(list_penjelasan_pasal_cleaned)):
  list_penjelasan_pasal_cleaned[i] = "[Berikut adalah penjelasan atas Pasal " + str(i) + " pada " + judul_UU_13_2003 + "] " + list_penjelasan_pasal_cleaned[i]

len(list_penjelasan_pasal_cleaned)

194

In [ ]:
# Memeriksa isi dari list pasal yang telah dibuat
list_penjelasan_pasal_cleaned[0]

'[Berikut adalah penjelasan atas Pasal 0 pada UU Nomor 13 Tahun 2003] PENJELASAN ATAS UNDANG -UNDANG REPUBLIK INDONESIA NOMOR 13 TAHUN 2003 TENTANG KETENAGAKERJAAN I.UMUM Pembangunan ketenagakerjaan sebagai bagian integral dari pembangunan nasional berdasarkan Pancasila dan Undang -undang Dasar Negara Republik Indonesia Tahun 1945, dilak sanakan dalam rangka pembangunan manusia Indonesia seutuhnya dan pembangunan masyarakat Indonesia seluruhnya untuk meningkatkan harkat, martabat, dan harga diri tenaga kerja serta mewujudkan masyarakat sejahtera, adil, makmur, dan merata, baik materiil mau pun spiritual. Pembangunan ketenagakerjaan harus diatur sedemikian rupa sehingga terpenuhi hak -hak dan perlindungan yang mendasar bagi tenaga kerja dan pekerja/buruh serta pada saat yang bersamaan dapat mewujudkan kondisi yang kondusif bagi pengembangan dunia usaha. Pembangunan ketenagakerjaan mempunyai banyak dimensi dan keterkaitan. Keterkaitan itu tidak hanya dengan kepentingan tenaga kerja selam

In [ ]:
# Memeriksa isi dari list pasal yang telah dibuat
list_penjelasan_pasal_cleaned[1]

'[Berikut adalah penjelasan atas Pasal 1 pada UU Nomor 13 Tahun 2003] Cukup jelas'

In [ ]:
# Memeriksa isi dari list pasal yang telah dibuat
list_penjelasan_pasal_cleaned[2]

'[Berikut adalah penjelasan atas Pasal 2 pada UU Nomor 13 Tahun 2003] Pembangunan ketenagakerjaan dilaksanakan dalam rangka pembangunan manusia Indonesia seutuhnya. Oleh sebab itu, pembangunan ketenaga -kerjaan dilaksanakan untuk mewujudkan manusia dan masyarakat Indonesia yang sejahtera, adil, makmur, dan merata baik materii l maupun spiritual.'

In [ ]:
# Memuat isi penjelasan atas pasal 1 sampai dengan 193 ke dalam list yang telah diinisialisasi di awal
# list_penjelasan_pasal_cleaned[1:] --> sebab indeks ke-0 dikecualikan
UU_13_2003 = UU_13_2003 + list_penjelasan_pasal_cleaned[1:]
len(UU_13_2003)

418

## Data Loading

In [ ]:
# Menuliskan data ke dalam file .txt supaya dapat dipergunakan untuk keperluan lain
with open('list_of_data_UU_13_2003.txt', 'w') as outfile:
  outfile.writelines(["{}\n".format(item) for item in UU_13_2003])

# UU 11/2020 tentang Cipta Kerja
Bab IV Bagian Kedua

In [ ]:
# Menginisialisasi list yang akan mengandung seluruh potongan peraturan
UU_11_2020 = []

## Text Extraction

In [ ]:
# Memilih sumber data (dokumen pdf)
file_name = os.path.basename('/content/1771 (1) (1).pdf')
judul_UU_11_2020 = "UU Nomor 11 Tahun 2020"
judul_UU_11_2020

'UU Nomor 11 Tahun 2020'

In [ ]:
# Dimodifikasi dari https://pypdf2.readthedocs.io/en/latest/user/extract-text.html
# Mengekstrak teks dari suatu dokumen
extracted_pdf_with_pypdf = ""
reader = PdfReader(file_name)

# Helper function untuk mengabaikan header dan footer
def visitor_body(text, cm, tm, fontDict, fontSize):
    y = tm[5]
    if y > 85 and y < 700: # dari bawah, lebih dari 85 dan tidak sampai 700
        parsed_parts.append(text)

for i in range(len(reader.pages)):
  parsed_parts = []
  reader.pages[i].extract_text(visitor_text=visitor_body)
  whole_page = "".join(parsed_parts)
  extracted_pdf_with_pypdf += whole_page

print(extracted_pdf_with_pypdf)


Bagian Kedua
Ketenagakerjaan
Pasal 81
Beberapa ketentuan dalam Undang-Undang Nomor 13 Tahun
2OO3 tentang Ketenagakerjaan (Lembaran Negara Republik
Indonesia Tahun 2OO3 Nomor 39, Tambahan Lembaran Negara
Republik Indonesia Nomor 42791diubah sebagai berikut:
1. Ketentuan Pasal 13 diubah sehingga berbunyi sebagai
berikut:
Pasal 13
(1) Pelatihan kerja diselenggarakan oleh:
a- lembaga pelatihan kerja pemerintah;
b. lembaga pelatihan kerja swasta; atau
c. lembaga pelatihan kerja perusahaan.
(21 Pelatihan kerja dapat diselenggarakan di tempat
pelatihan atau tempat kerja.
(3) Lembaga pelatihan kerja pemerintah sebagaimana
dimaksud pada ayat (1) huruf a dalam
menyelenggarakan pelatihan kerja dapat bekerja
sama dengan swasta.
(41 Lembaga pelatihan kerja pemerintah sebagaimana
dimaksud pada ayat (1) huruf a dan lembaga
pelatihan kerja perusahaan sebagaimana dimaksud
pada ayat (1) huruf c mendaftarkan kegiatannya
kepada instansi yang bertanggung jawab di bidang
ketenagakerjaan di kabupaten/ kota.

## Text Preprocessing

In [ ]:
extracted_pdf_with_pypdf

'\nBagian Kedua\nKetenagakerjaan\nPasal 81\nBeberapa ketentuan dalam Undang-Undang Nomor 13 Tahun\n2OO3 tentang Ketenagakerjaan (Lembaran Negara Republik\nIndonesia Tahun 2OO3 Nomor 39, Tambahan Lembaran Negara\nRepublik Indonesia Nomor 42791diubah sebagai berikut:\n1. Ketentuan Pasal 13 diubah sehingga berbunyi sebagai\nberikut:\nPasal 13\n(1) Pelatihan kerja diselenggarakan oleh:\na- lembaga pelatihan kerja pemerintah;\nb. lembaga pelatihan kerja swasta; atau\nc. lembaga pelatihan kerja perusahaan.\n(21 Pelatihan kerja dapat diselenggarakan di tempat\npelatihan atau tempat kerja.\n(3) Lembaga pelatihan kerja pemerintah sebagaimana\ndimaksud pada ayat (1) huruf a dalam\nmenyelenggarakan pelatihan kerja dapat bekerja\nsama dengan swasta.\n(41 Lembaga pelatihan kerja pemerintah sebagaimana\ndimaksud pada ayat (1) huruf a dan lembaga\npelatihan kerja perusahaan sebagaimana dimaksud\npada ayat (1) huruf c mendaftarkan kegiatannya\nkepada instansi yang bertanggung jawab di bidang\nketenaga

In [ ]:
# Mempartisi bab iv bagian kedua dari bagian ketiga
partioned_string_1, partioned_string_2, partioned_string_3  = extracted_pdf_with_pypdf.partition("\nBagian Ketiga")
partioned_string_1

'\nBagian Kedua\nKetenagakerjaan\nPasal 81\nBeberapa ketentuan dalam Undang-Undang Nomor 13 Tahun\n2OO3 tentang Ketenagakerjaan (Lembaran Negara Republik\nIndonesia Tahun 2OO3 Nomor 39, Tambahan Lembaran Negara\nRepublik Indonesia Nomor 42791diubah sebagai berikut:\n1. Ketentuan Pasal 13 diubah sehingga berbunyi sebagai\nberikut:\nPasal 13\n(1) Pelatihan kerja diselenggarakan oleh:\na- lembaga pelatihan kerja pemerintah;\nb. lembaga pelatihan kerja swasta; atau\nc. lembaga pelatihan kerja perusahaan.\n(21 Pelatihan kerja dapat diselenggarakan di tempat\npelatihan atau tempat kerja.\n(3) Lembaga pelatihan kerja pemerintah sebagaimana\ndimaksud pada ayat (1) huruf a dalam\nmenyelenggarakan pelatihan kerja dapat bekerja\nsama dengan swasta.\n(41 Lembaga pelatihan kerja pemerintah sebagaimana\ndimaksud pada ayat (1) huruf a dan lembaga\npelatihan kerja perusahaan sebagaimana dimaksud\npada ayat (1) huruf c mendaftarkan kegiatannya\nkepada instansi yang bertanggung jawab di bidang\nketenaga

In [ ]:
# Mempartisi nomor pasal
partioned_string_1, partioned_string_2, partioned_string_3  = partioned_string_1.partition("42791diubah sebagai berikut:")
bagian_kedua = partioned_string_3
bagian_kedua

'\n1. Ketentuan Pasal 13 diubah sehingga berbunyi sebagai\nberikut:\nPasal 13\n(1) Pelatihan kerja diselenggarakan oleh:\na- lembaga pelatihan kerja pemerintah;\nb. lembaga pelatihan kerja swasta; atau\nc. lembaga pelatihan kerja perusahaan.\n(21 Pelatihan kerja dapat diselenggarakan di tempat\npelatihan atau tempat kerja.\n(3) Lembaga pelatihan kerja pemerintah sebagaimana\ndimaksud pada ayat (1) huruf a dalam\nmenyelenggarakan pelatihan kerja dapat bekerja\nsama dengan swasta.\n(41 Lembaga pelatihan kerja pemerintah sebagaimana\ndimaksud pada ayat (1) huruf a dan lembaga\npelatihan kerja perusahaan sebagaimana dimaksud\npada ayat (1) huruf c mendaftarkan kegiatannya\nkepada instansi yang bertanggung jawab di bidang\nketenagakerjaan di kabupaten/ kota.\nKetentuan Pasal\nberikut:14 diubah sehingga berbunyi sebagai\nPasal 14\n(1) Lembaga pelatihan kerja swasta sebagaimana\ndimaksud dalam Pasal 13 ayat (1) huruf b wajib\nmemenuhi Pertzinan Berusaha yang diterbitkan oleh2\n(2) Bagi lembag

In [ ]:
# Mengambil pasal yang dihapus
pattern = "Pasal \d+ dihapus\.?"
list_pasal_dihapus = re.findall(pattern, bagian_kedua)
len(list_pasal_dihapus)

29

In [ ]:
# Menghilangkan pasal yang dihapus
b2_tanpa_hapus = re.sub(pattern, '', bagian_kedua)
len(b2_tanpa_hapus)

36134

In [ ]:
b2_tanpa_hapus

'\n1. Ketentuan Pasal 13 diubah sehingga berbunyi sebagai\nberikut:\nPasal 13\n(1) Pelatihan kerja diselenggarakan oleh:\na- lembaga pelatihan kerja pemerintah;\nb. lembaga pelatihan kerja swasta; atau\nc. lembaga pelatihan kerja perusahaan.\n(21 Pelatihan kerja dapat diselenggarakan di tempat\npelatihan atau tempat kerja.\n(3) Lembaga pelatihan kerja pemerintah sebagaimana\ndimaksud pada ayat (1) huruf a dalam\nmenyelenggarakan pelatihan kerja dapat bekerja\nsama dengan swasta.\n(41 Lembaga pelatihan kerja pemerintah sebagaimana\ndimaksud pada ayat (1) huruf a dan lembaga\npelatihan kerja perusahaan sebagaimana dimaksud\npada ayat (1) huruf c mendaftarkan kegiatannya\nkepada instansi yang bertanggung jawab di bidang\nketenagakerjaan di kabupaten/ kota.\nKetentuan Pasal\nberikut:14 diubah sehingga berbunyi sebagai\nPasal 14\n(1) Lembaga pelatihan kerja swasta sebagaimana\ndimaksud dalam Pasal 13 ayat (1) huruf b wajib\nmemenuhi Pertzinan Berusaha yang diterbitkan oleh2\n(2) Bagi lembag

In [ ]:
# Fungsi untuk menghilangkan bagian yang tidak diinginkan setelah parsing
def text_preprocessing_1(input):
  temp = re.sub('Ketentuan\. ?\. ?\. ?', '', input)
  temp = re.sub('Dalam\. ?\. ?\. ?', '', temp)
  temp = re.sub('Pasal(\s?(I57|\d{1,3}[A-Z]?)\s*)?\. ?\. ?\. ?', '', temp)
  temp = re.sub('berhalangan\. ?\. ?\. ?', '', temp)
  temp = re.sub('Pemutusan\. ?\. ?\. ?', '', temp)
  temp = re.sub('membujuk\. ?\. ?\. ?', '', temp)
  output = re.sub('pekerja\. ?\. ?\. ?', '', temp)
  return output

In [ ]:
# Mengurai keseluruhan teks berdasarkan nomor pasal yang ada (parsing)
pattern = "\\nPasal\s\d+[A-Z]?\\n"
list_nomor = re.findall(pattern, b2_tanpa_hapus)

list_pasal = [result for result in re.split(pattern, text_preprocessing_1(b2_tanpa_hapus))][1:]

In [ ]:
# Mengurai pasal yang belum terurai
splitters = ["\nPasal TT\n", "Pasal 88B\n", "Pasal 88C\n", "Pasal 95\n", "Pasal 154A\n"]
result_flat = list_pasal

for i in range(len(splitters)):
  result = [item.split(splitters[i]) for item in result_flat]
  result_flat = [item for l in result for item in l]

In [ ]:
print(len(result_flat))
result_flat

44


['(1) Pelatihan kerja diselenggarakan oleh:\na- lembaga pelatihan kerja pemerintah;\nb. lembaga pelatihan kerja swasta; atau\nc. lembaga pelatihan kerja perusahaan.\n(21 Pelatihan kerja dapat diselenggarakan di tempat\npelatihan atau tempat kerja.\n(3) Lembaga pelatihan kerja pemerintah sebagaimana\ndimaksud pada ayat (1) huruf a dalam\nmenyelenggarakan pelatihan kerja dapat bekerja\nsama dengan swasta.\n(41 Lembaga pelatihan kerja pemerintah sebagaimana\ndimaksud pada ayat (1) huruf a dan lembaga\npelatihan kerja perusahaan sebagaimana dimaksud\npada ayat (1) huruf c mendaftarkan kegiatannya\nkepada instansi yang bertanggung jawab di bidang\nketenagakerjaan di kabupaten/ kota.\nKetentuan Pasal\nberikut:14 diubah sehingga berbunyi sebagai',
 '(1) Lembaga pelatihan kerja swasta sebagaimana\ndimaksud dalam Pasal 13 ayat (1) huruf b wajib\nmemenuhi Pertzinan Berusaha yang diterbitkan oleh2\n(2) Bagi lembaga pelatihan kerja swasta yang terdapat\npenyertaan modal asing, Perizinan Berusaha\n

In [ ]:
# Menambahkan list nomor
nomor_pasal_setelah = ["\nPasal 78\n", "\nPasal 88D\n", "\nPasal 88D\n", "\nPasal 98\n", "\nPasal 156\n"]
for i in range(len(splitters)):
  list_nomor.insert(list_nomor.index(nomor_pasal_setelah[i]), splitters[i])

# Mengubah nilai "Pasal TT" menjadi "Pasal 77"
list_nomor[14] = "\nPasal 77\n"

# Membersihkan nomor pasal
for i in range(len(list_nomor)):
  list_nomor[i] = re.findall("\d+[A-Z]?", list_nomor[i])[0]

for i in range(len(list_pasal_dihapus)):
  list_pasal_dihapus[i] = re.findall("\d+[A-Z]?", list_pasal_dihapus[i])[0]

In [ ]:
print(len(list_pasal_dihapus))
list_pasal_dihapus

29


['43',
 '44',
 '46',
 '48',
 '64',
 '65',
 '89',
 '90',
 '91',
 '96',
 '97',
 '152',
 '154',
 '155',
 '158',
 '159',
 '161',
 '162',
 '163',
 '164',
 '165',
 '166',
 '167',
 '168',
 '169',
 '170',
 '171',
 '172',
 '184']

In [ ]:
print(len(list_nomor))
list_nomor

44


['13',
 '14',
 '37',
 '42',
 '45',
 '47',
 '49',
 '56',
 '57',
 '58',
 '59',
 '61',
 '61A',
 '66',
 '77',
 '78',
 '79',
 '88',
 '88A',
 '88B',
 '88C',
 '88D',
 '88E',
 '90A',
 '90B',
 '92',
 '92A',
 '94',
 '95',
 '98',
 '151',
 '151A',
 '153',
 '154A',
 '156',
 '157',
 '157A',
 '160',
 '185',
 '186',
 '187',
 '188',
 '190',
 '191A']

In [ ]:
# Fungsi untuk menghilangkan bagian yang tidak diinginkan setelah parsing
def text_preprocessing_2(input):
  temp = re.sub('Ketentuan Pasal.*', '', input)
  output = re.sub('Di antara Pasal.*', '', temp)
  return output

In [ ]:
# Menghilangkan semua karakter setelah frase "Ketentuan Pasal" dan "Di antara Pasal"
list_pasal_cleaned = []

for i in range(len(result_flat)):
  partioned_string_1, partioned_string_2, partioned_string_3  = result_flat[i].partition("Ketentuan Pasal")
  list_pasal_cleaned.append(partioned_string_1)

for i in range(len(list_pasal_cleaned)):
  partioned_string_1, partioned_string_2, partioned_string_3  = list_pasal_cleaned[i].partition("Di antara Pasal")
  list_pasal_cleaned[i] = partioned_string_1

In [ ]:
print(len(list_pasal_cleaned))
list_pasal_cleaned

44


['(1) Pelatihan kerja diselenggarakan oleh:\na- lembaga pelatihan kerja pemerintah;\nb. lembaga pelatihan kerja swasta; atau\nc. lembaga pelatihan kerja perusahaan.\n(21 Pelatihan kerja dapat diselenggarakan di tempat\npelatihan atau tempat kerja.\n(3) Lembaga pelatihan kerja pemerintah sebagaimana\ndimaksud pada ayat (1) huruf a dalam\nmenyelenggarakan pelatihan kerja dapat bekerja\nsama dengan swasta.\n(41 Lembaga pelatihan kerja pemerintah sebagaimana\ndimaksud pada ayat (1) huruf a dan lembaga\npelatihan kerja perusahaan sebagaimana dimaksud\npada ayat (1) huruf c mendaftarkan kegiatannya\nkepada instansi yang bertanggung jawab di bidang\nketenagakerjaan di kabupaten/ kota.\n',
 '(1) Lembaga pelatihan kerja swasta sebagaimana\ndimaksud dalam Pasal 13 ayat (1) huruf b wajib\nmemenuhi Pertzinan Berusaha yang diterbitkan oleh2\n(2) Bagi lembaga pelatihan kerja swasta yang terdapat\npenyertaan modal asing, Perizinan Berusaha\nsebagaimana dimaksud pada ayat (1) diterbitkan oleh\nPemerin

In [ ]:
# Memodifikasi teks (menambahkan identifier tertentu)
# Harapannya suatu LLM dapat diinstruksikan untuk mengenali isi dari tiap nomor pasal berdasarkan identifier tersebut
mod_list_pasal = []
for i in range(len(list_pasal_cleaned)):
  nomor_pasal = list_nomor[i]
  isi_pasal = list_pasal_cleaned[i]
  is_baru = bool(re.search("\d+[A-Z]", nomor_pasal))
  if (is_baru):
    mod_list_pasal.append("[Berikut isi Pasal " + nomor_pasal + " UU Nomor 13 Tahun 2003" + " yang ditambahkan oleh " + judul_UU_11_2020 + "] " + isi_pasal)
  else:
    mod_list_pasal.append("[Berikut isi Pasal " + nomor_pasal + " pada UU Nomor 13 Tahun 2003 yang diubah oleh " + judul_UU_11_2020 + "] " + isi_pasal)

In [ ]:
print(len(mod_list_pasal))
mod_list_pasal

44


['[Berikut isi Pasal 13 pada UU Nomor 13 Tahun 2003 yang diubah oleh UU Nomor 11 Tahun 2020] (1) Pelatihan kerja diselenggarakan oleh:\na- lembaga pelatihan kerja pemerintah;\nb. lembaga pelatihan kerja swasta; atau\nc. lembaga pelatihan kerja perusahaan.\n(21 Pelatihan kerja dapat diselenggarakan di tempat\npelatihan atau tempat kerja.\n(3) Lembaga pelatihan kerja pemerintah sebagaimana\ndimaksud pada ayat (1) huruf a dalam\nmenyelenggarakan pelatihan kerja dapat bekerja\nsama dengan swasta.\n(41 Lembaga pelatihan kerja pemerintah sebagaimana\ndimaksud pada ayat (1) huruf a dan lembaga\npelatihan kerja perusahaan sebagaimana dimaksud\npada ayat (1) huruf c mendaftarkan kegiatannya\nkepada instansi yang bertanggung jawab di bidang\nketenagakerjaan di kabupaten/ kota.\n',
 '[Berikut isi Pasal 14 pada UU Nomor 13 Tahun 2003 yang diubah oleh UU Nomor 11 Tahun 2020] (1) Lembaga pelatihan kerja swasta sebagaimana\ndimaksud dalam Pasal 13 ayat (1) huruf b wajib\nmemenuhi Pertzinan Berusaha y

In [ ]:
# Fungsi untuk menghilangkan bagian yang tidak diinginkan setelah parsing
def text_preprocessing_2(input):
  temp = re.sub('\\n', ' ', input)
  return temp

In [ ]:
# Menghilangkan bagian yang tidak diinginkan setelah parsing
list_pasal_cleaned_2 = []
for i in range(len(mod_list_pasal)):
  list_pasal_cleaned_2.append(text_preprocessing_2(mod_list_pasal[i]))

len(list_pasal_cleaned_2)
UU_11_2020 = list_pasal_cleaned_2

In [ ]:
# Menambahkan pasal yang dihapus
temp = "[Berikut {} buah pasal dari UU Nomor 13 Tahun 2003 yang telah dihapus pada UU Nomor 11 Tahun 2020]".format(len(list_pasal_dihapus))
for pasal in list_pasal_dihapus:
  temp += " Pasal {},".format(pasal)

UU_11_2020.append(temp)

In [ ]:
print(len(UU_11_2020))
UU_11_2020

45


['[Berikut isi Pasal 13 pada UU Nomor 13 Tahun 2003 yang diubah oleh UU Nomor 11 Tahun 2020] (1) Pelatihan kerja diselenggarakan oleh: a- lembaga pelatihan kerja pemerintah; b. lembaga pelatihan kerja swasta; atau c. lembaga pelatihan kerja perusahaan. (21 Pelatihan kerja dapat diselenggarakan di tempat pelatihan atau tempat kerja. (3) Lembaga pelatihan kerja pemerintah sebagaimana dimaksud pada ayat (1) huruf a dalam menyelenggarakan pelatihan kerja dapat bekerja sama dengan swasta. (41 Lembaga pelatihan kerja pemerintah sebagaimana dimaksud pada ayat (1) huruf a dan lembaga pelatihan kerja perusahaan sebagaimana dimaksud pada ayat (1) huruf c mendaftarkan kegiatannya kepada instansi yang bertanggung jawab di bidang ketenagakerjaan di kabupaten/ kota. ',
 '[Berikut isi Pasal 14 pada UU Nomor 13 Tahun 2003 yang diubah oleh UU Nomor 11 Tahun 2020] (1) Lembaga pelatihan kerja swasta sebagaimana dimaksud dalam Pasal 13 ayat (1) huruf b wajib memenuhi Pertzinan Berusaha yang diterbitkan ol

## Data Loading

In [ ]:
# Menuliskan data ke dalam file .txt supaya dapat dipergunakan untuk keperluan lain
with open('list_of_data_UU_11_2020.txt', 'w') as outfile:
  outfile.writelines(["{}\n".format(item) for item in UU_11_2020])

# PP 34/2021 tentang Penggunaan Tenaga Kerja Asing

## Text Processing

In [ ]:
# Menginisialisasi list yang akan mengandung seluruh potongan peraturan
# PP 34/2021 berisikan 48 pasal
PP_34_2021 = []

In [ ]:
# Mengambil data dari sumber data
file_name = os.path.basename('/content/PERATURAN PEMERINTAH REPUBLIK INDONESIA NOMOR 34 TAHUN 2021.txt')
print(file_name)
with open(file_name) as f:
    txt = f.read()

PERATURAN PEMERINTAH REPUBLIK INDONESIA NOMOR 34 TAHUN 2021.txt


In [ ]:
# Mengurai keseluruhan teks berdasarkan nomor pasal yang ada (parsing)
seperator = "### [Pasal"
daftar_pasal =  [seperator+result for result in txt.split(seperator)]
len(daftar_pasal)

49

In [ ]:
# Memeriksa isi dari list pasal yang telah dibuat
daftar_pasal[0]

'### [Pasal# [PERATURAN PEMERINTAH REPUBLIK INDONESIA TAHUN 2021 NOMOR 34](http://example.org/legal/peraturan/pp/2021/34)\n\n| Nama | Data |\n| ------ | ----- |\n|Name|PERATURAN PEMERINTAH REPUBLIK INDONESIA TAHUN 2021 NOMOR 34|\n|Tentang| PENGGUNAAN TENAGA KERJA ASING|\n|Memutuskan||\n|Tempat Disahkan|Jakarta|\n|Tanggal Disahkan|2|\n|Bulan Disahkan|2|\n|Tahun Disahkan|2021|\n|Jabatan Pengesah|PRESIDEN REPUBLIK INDONESIA,|\n|Nama Pengesah|JOKO WIDODO|\n|Dengan Persetujuan||\n# [Menimbang](http://example.org/legal/peraturan/pp/2021/34/menimbang)\nbahwa untuk melaksanakan ketentuan Pasal 81 dan Pasal 185 huruf b [Undang-Undang Nomor 11 Tahun 2020](http://example.org/legal/peraturan/uu/2020/11) tentang Cipta Kerja, perlu menetapkan Peraturan Pemerintah tentang Penggunaan Tenaga Kerja Asing,\n# [Mengingat](http://example.org/legal/peraturan/pp/2021/34/mengingat)\n\n* [1.](http://example.org/legal/peraturan/pp/2021/34/mengingat/huruf/0001) Pasal 5 ayat (2) [Undang-Undang Dasar Negara Republ

In [ ]:
# Memeriksa isi dari list pasal yang telah dibuat
daftar_pasal[48]

'### [Pasal 48](http://example.org/legal/peraturan/pp/2021/34/pasal/0048)\nPeraturan Pemerintah ini mulai berlaku pada tanggal 1 April 2021. Agar setiap orang mengetahuinya, memerintahkan pengundangan Peraturan Pemerintah ini dengan penempatannya dalam Lembaran Negara Republik Indonesia.\n \n'

In [ ]:
def article_text_processing(pasal, judul_peraturan):
  # Menghilangkan bagian yang tidak diinginkan setelah parsing
  text = re.sub(r'#\s\[BAB.*', '', pasal)
  text = re.sub(r'##\s\[Bagian.*', '', text)
  text = re.sub(r'##\s\[Paragraf.*', '', text)
  text = re.sub(r'\(http[s]?://\S+', '', text)
  text = re.sub(r'[*#]', '', text)
  text = re.sub(r'\n', ' ', text)

  # Memodifikasi teks (menambahkan identifier tertentu)
  # Harapannya suatu LLM dapat diinstruksikan untuk mengenali isi dari tiap nomor pasal berdasarkan identifier tersebut
  temp = re.findall(r'\[Pasal \d*\]', text)[0]
  pasal_id = re.sub(r'[\[\]]', '', temp)
  text = re.sub(r'\[Pasal \d*\]', '', text)
  identifier = "[Berikut adalah isi {} pada {}]".format(pasal_id, judul_peraturan)
  text = identifier + " " + text

  # Menghilangkan line break yang terlalu banyak
  text = re.sub(' +', ' ', text.replace('\n', ' '))
  text = text.strip()
  return text

In [ ]:
for i in range(1, len(daftar_pasal)):
  PP_34_2021.append(article_text_processing(daftar_pasal[i], "PP Nomor 34 Tahun 2021"))

In [ ]:
# Memeriksa isi dari list pasal yang telah dibuat
PP_34_2021[0]

'[Berikut adalah isi Pasal 1 pada PP Nomor 34 Tahun 2021] Dalam Peraturan Pemerintah ini yang dimaksud dengan: [1.] Tenaga Kerja Asing yang selanjutnya disingkat TKA adalah warga negara asing pemegang visa dengan maksud bekerja di wilayah Indonesia. [2.] Pemberi Kerja TKA adalah badan hukum yang didirikan berdasarkan hukum Indonesia atau badan lannya yang mempekerjakan TKA dengan membayar upah atau imbalan dalam bentuk lain. [3.] Tenaga Kerja Pendamping TKA adalah tenaga kerja Indonesia yang ditunjuk oleh Pemberi Kerja TKA dan dipersiapkan sebagai pendamping TKA yang dipekerjakan dalam rangka alih teknologi dan alih keahlian. [4.] Rencana Penggunaan Tenaga Kerja Asing yang selanjutnya disingkat RPTKA adalah rencana penggunaan TKA pada jabatan tertentu dan jangka waktu tertentu. [5.] Pengesahan Rencana Penggunaan Tenaga Kerja Asing yang selanjutnya disebut Pengesahan RPTKA adalah persetujuan penggunaan TKA yang disahkan oleh menteri yang menyelenggarakan urusan pemerintahan di bidang ke

**Memecah Pasal 1 sebab mengandung terlalu banyak konteks berbeda.**

In [ ]:
pasal_1_splitted = re.split('\[\d+\.\]\s', PP_34_2021[0])
pasal_1_splitted[1]

'Tenaga Kerja Asing yang selanjutnya disingkat TKA adalah warga negara asing pemegang visa dengan maksud bekerja di wilayah Indonesia. '

In [ ]:
# Memodifikasi teks (menambahkan identifier tertentu)
# Harapannya suatu LLM dapat diinstruksikan untuk mengenali isi dari tiap nomor pasal berdasarkan identifier tersebut
pasal_1_modified = []

for i in range(1, len(pasal_1_splitted)):
  pasal_1_modified.append("[Berikut adalah isi Pasal 1 angka " + str(i) + " pada " + "PP 34/2021" + "] " + pasal_1_splitted[i])


In [ ]:
pasal_1_modified[0]

'[Berikut adalah isi Pasal 1 angka 1 pada PP 34/2021] Tenaga Kerja Asing yang selanjutnya disingkat TKA adalah warga negara asing pemegang visa dengan maksud bekerja di wilayah Indonesia. '

**Menggabungkan isi pasal 1 dengan pasal-pasal lainnya ke dalam satu list**

In [ ]:
FINAL_PP_34_2021 = pasal_1_modified + PP_34_2021[1:]
FINAL_PP_34_2021

['[Berikut adalah isi Pasal 1 angka 1 pada PP 34/2021] Tenaga Kerja Asing yang selanjutnya disingkat TKA adalah warga negara asing pemegang visa dengan maksud bekerja di wilayah Indonesia. ',
 '[Berikut adalah isi Pasal 1 angka 2 pada PP 34/2021] Pemberi Kerja TKA adalah badan hukum yang didirikan berdasarkan hukum Indonesia atau badan lannya yang mempekerjakan TKA dengan membayar upah atau imbalan dalam bentuk lain. ',
 '[Berikut adalah isi Pasal 1 angka 3 pada PP 34/2021] Tenaga Kerja Pendamping TKA adalah tenaga kerja Indonesia yang ditunjuk oleh Pemberi Kerja TKA dan dipersiapkan sebagai pendamping TKA yang dipekerjakan dalam rangka alih teknologi dan alih keahlian. ',
 '[Berikut adalah isi Pasal 1 angka 4 pada PP 34/2021] Rencana Penggunaan Tenaga Kerja Asing yang selanjutnya disingkat RPTKA adalah rencana penggunaan TKA pada jabatan tertentu dan jangka waktu tertentu. ',
 '[Berikut adalah isi Pasal 1 angka 5 pada PP 34/2021] Pengesahan Rencana Penggunaan Tenaga Kerja Asing yang s

## Data Loading

In [ ]:
# Menuliskan data ke dalam file .txt supaya dapat dipergunakan untuk keperluan lain
with open('list_of_data_PP_34_2021.txt', 'w') as outfile:
  outfile.writelines(["{}\n".format(item) for item in FINAL_PP_34_2021])

# PP 37/2021 tentang Penyelenggaraan Program Jaminan Kehilangan Pekerjaan

## Text Extraction

In [ ]:
# Memilih sumber data (dokumen pdf)
file_name = os.path.basename('/content/PP Nomor 37 Tahun 2021.pdf')

In [ ]:
# Dimodifikasi dari https://pypdf2.readthedocs.io/en/latest/user/extract-text.html
# Mengekstrak teks dari suatu dokumen
extracted_pdf_with_pypdf = ""
reader = PdfReader(file_name)

# Helper function untuk mengabaikan header dan footer
def visitor_body(text, cm, tm, fontDict, fontSize):
    y = tm[5]
    if y > 70 and y < 840: # dari bawah, lebih dari 70 dan tidak sampai 840
        parsed_parts.append(text)

for i in range(len(reader.pages)):
  parsed_parts = []
  reader.pages[i].extract_text(visitor_text=visitor_body)
  whole_page = "".join(parsed_parts)
  extracted_pdf_with_pypdf += whole_page

print(extracted_pdf_with_pypdf)

 
 
LEMBARAN NEGARA  
REPUBLIK INDONESIA  
No.47, 2021  KESRA. Program Jaminan Kehilangan 
Pekerjaan . Penyelenggaraan . (Penjelasan dalam 
Tambahan Lembaran Negara Republik 
Indonesia Nomor 6 649) 
 
PERATURAN PEMERINTAH REPUBLIK INDONESIA  
NOMO R 37 TAHUN 2021 
TENTANG  
PENYELENGGARAAN PROGRAM JAMINAN KEHILANGAN PEKERJAAN  
 
DENGAN RAHMAT TUHAN YANG MAHA ESA  
 
PRESIDEN REPUBLIK INDONESIA,  
 
 
Menimbang  : bahwa  untuk melaksanakan ketentuan Pasal 82 dan Pasal 
185 huruf b Undang -Undang Nomor 11 Tahun 2020 
tentang Cipta Kerja , perlu menetapkan Peraturan 
Pemerintah tentang Penyelenggaraan Program Jaminan 
Kehilangan Pekerjaan;  
 
Mengingat  : 1.  Pasal 5 ayat (2) Undang -Undang Dasar Negara  
Republik Indonesia Tahun 1945;  
2. Undang -Undang Nomor 13 Tahun 2003 tentang 
Ketenagakerjaan ( Lembaran Negara Republik 
Indonesia Tahun 20 03 Nomor 39, Tambahan 
Lembaran Negara Republik Indonesia Nomor 4279);  
3. Undang -Undang Nomor 40 Tahun 2004  tentang 
Sistem Jaminan Sosial 

## Text Preprocessing

In [ ]:
extracted_pdf_with_pypdf

In [ ]:
# Mempartisi bagian penjelasan
partioned_string_1, partioned_string_2, partioned_string_3  = extracted_pdf_with_pypdf.partition("\nPENJELASAN  \nATAS")
bagian_penjelasan = partioned_string_2 + partioned_string_3
bagian_penjelasan

In [ ]:
# Mempartisi bagian penutup
partioned_string_1, partioned_string_2, partioned_string_3 = partioned_string_1.partition("Ditetapkan d i")
bagian_penutup = partioned_string_2 + partioned_string_3
bagian_penutup

'Ditetapkan d i Jakarta  \npada tanggal 2 Februari 2021  \n \nPRESIDEN REPUBLIK INDONESIA,  \n \nttd. \n \nJOKO WIDODO  \n \nDiundangkan di Jakarta  \npada tanggal 2 Februari 2021  \n \nMENTERI HUKUM DAN HAK ASASI MANUSIA  \nREPUBLIK INDONESIA,  \n \nttd. \n \nYASONNA H. LAOLY  \n \n \nTAMBAHAN  \nLEMBARAN NEGARA R .I \nNo.6649 KESRA. Program Jaminan Kehilangan Pekerjaan . \nPenyelenggaraan . (Penjelasan atas Lembaran \nNegara Republik Indonesia Tahun 202 1 Nomor 47) \n \n '

In [ ]:
# Mempartisi bagian batang tubuh
partioned_string_1, partioned_string_2, partioned_string_3 = partioned_string_1.partition("BAB I \nKETENTUAN UMUM")
bagian_batang_tubuh = partioned_string_2 + partioned_string_3
bagian_batang_tubuh

'BAB I \nKETENTUAN UMUM  \n \nPasal 1  \nDalam Peraturan Pemerintah ini yang dimaksud dengan:  \n1. Jaminan Kehilangan Pekerjaan yang selanjutnya \ndisingkat JKP adalah jamina n sosial  yang diberikan \nkepada Pekerja/Buruh yang mengalami Pemutusan \nHubungan  Kerja berupa manfaat uang tunai , akses \ninformasi pasar kerja, dan Pelatihan Kerja.  \n2. Pekerja/Buruh adalah setiap orang yang bekerja \ndengan menerima Upah atau imbalan dalam bentuk \nlain. \n3. Pengusaha adalah:  \na. orang perseorangan, persekutuan, atau badan \nhukum yan g menjalankan suatu perusahaan \nmilik sendiri;  \nb. orang perseorangan, persekutuan, atau badan \nhukum yang secara berdiri sendiri menjalankan \nperusahaan bukan miliknya;   \nc. orang perseorangan, persekutuan, atau badan \nhukum yang berada di Indonesia mewakili \nperusaha an sebagaimana dimaksud dalam huruf \na dan huruf b yang berkedudukan di luar wilayah \n\nIndonesia.  \n4. Upah adalah hak Pekerja/Buruh yang diterima dan \ndinyatakan dalam bentu

In [ ]:
# Mengurai keseluruhan teks berdasarkan nomor pasal yang ada (parsing)
pattern = "\\nPasal\s\d+\s\s\\n|\\nPasal\s\s\d+\s\\n|\\nPasal\s\s\d+\s\s\\n|\\nPasal\s\d+\s\\n|\\nPasal\s\d\s\d\s\\n"
list_pasal = []
list_pasal =  [result for result in re.split(pattern, bagian_batang_tubuh)]

In [ ]:
# Seharusnya terdapat 50 pasal
# Entri pertama pada list di bawah bukanlah pasal
len(list_pasal)

51

In [ ]:
# Entri kedua pada list adalah pasal ke-1
list_pasal[1]

'Dalam Peraturan Pemerintah ini yang dimaksud dengan:  \n1. Jaminan Kehilangan Pekerjaan yang selanjutnya \ndisingkat JKP adalah jamina n sosial  yang diberikan \nkepada Pekerja/Buruh yang mengalami Pemutusan \nHubungan  Kerja berupa manfaat uang tunai , akses \ninformasi pasar kerja, dan Pelatihan Kerja.  \n2. Pekerja/Buruh adalah setiap orang yang bekerja \ndengan menerima Upah atau imbalan dalam bentuk \nlain. \n3. Pengusaha adalah:  \na. orang perseorangan, persekutuan, atau badan \nhukum yan g menjalankan suatu perusahaan \nmilik sendiri;  \nb. orang perseorangan, persekutuan, atau badan \nhukum yang secara berdiri sendiri menjalankan \nperusahaan bukan miliknya;   \nc. orang perseorangan, persekutuan, atau badan \nhukum yang berada di Indonesia mewakili \nperusaha an sebagaimana dimaksud dalam huruf \na dan huruf b yang berkedudukan di luar wilayah \n\nIndonesia.  \n4. Upah adalah hak Pekerja/Buruh yang diterima dan \ndinyatakan dalam bentuk uang sebagai imbalan dari \nPengusaha 

### Pasal 1
Memecah Pasal 1 sebab mengandung terlalu banyak konteks berbeda.

In [ ]:
pasal_1_splitted = re.split('\\n\d+\.\s', list_pasal[1])
len(pasal_1_splitted)

20

In [ ]:
pasal_1_splitted[1]

'Jaminan Kehilangan Pekerjaan yang selanjutnya \ndisingkat JKP adalah jamina n sosial  yang diberikan \nkepada Pekerja/Buruh yang mengalami Pemutusan \nHubungan  Kerja berupa manfaat uang tunai , akses \ninformasi pasar kerja, dan Pelatihan Kerja.  '

In [ ]:
pasal_1_modified = []
for i in range(1, len(pasal_1_splitted)):
  # remove multiple \n and whitespace
  temp = re.sub(' +', ' ', pasal_1_splitted[i].replace('\n', ' ')).strip()
  # modify the text so that (hopefully) the LLM can be instructed to easily identify the article number
  pasal_1_modified.append("[Berikut adalah isi Pasal 1 angka " + str(i) + " pada " + "PP 37/2021" + "] " + temp)

In [ ]:
pasal_1_modified[18]

'[Berikut adalah isi Pasal 1 angka 19 pada PP 37/2021] Menteri adalah menteri yang menyelenggarakan urusan pem erintahan di bidang ketenagakerjaan.'

### Pasal 2-50

In [ ]:
# Entri ketiga pada list adalah pasal ke-2
list_pasal[2]

'(1) Pengusaha wajib mengikutsertakan Pekerja/Buruh \nsebagai Peserta  dalam program JKP . \n(2) Program  JKP sebagaimana dimaksud pada aya t (1) \ndiselenggarakan untuk mempertahankan derajat \nkehidupan yang layak pada saat Pekerja/Buruh \nkehilangan pekerjaan.  \n '

In [ ]:
# Entri keempat pada list adalah pasal ke-3
list_pasal[3]

'JKP sebagaimana dimaksud dalam Pasal 2 diselenggarakan \noleh BPJS  Ketenagakerjaan dan Pemerintah  Pusat .  \n \nBAB II  \nKEPESERTAAN DAN T ATA CARA PENDAFTARAN  \n \nBagian Kesatu  \nKepesertaan  \n '

> Bisa dilihat pada contoh di atas, bahwa terdapat bagian yang tidak diinginkan (unwanted parts) setelah proses parsing. Penanganannya akan dilakukan secara manual.

In [ ]:
'''
daftar entri yang mengandung unwanted parts:
(\nBAB II  \nKEPESERTAAN...)
- list_pasal[3]
- list_pasal[10]
- list_pasal[17]
- list_pasal[41]
- list_pasal[43]
- list_pasal[44]
- list_pasal[45]
- list_pasal[47]

(\nBagian Kedua  \nTata Cara...)
- list_pasal[4]
- list_pasal[13]
- list_pasal[20]
- list_pasal[24]
- list_pasal[29]
- list_pasal[34]
'''

'\ndaftar entri yang mengandung unwanted parts:\n(\nBAB II  \nKEPESERTAAN...)\n- list_pasal[3]\n- list_pasal[10]\n- list_pasal[17]\n- list_pasal[41] \n- list_pasal[43]\n- list_pasal[44]\n- list_pasal[45]\n- list_pasal[47]\n\n(\nBagian Kedua  \nTata Cara...)\n- list_pasal[4]\n- list_pasal[13]\n- list_pasal[20]\n- list_pasal[24]\n- list_pasal[29]\n- list_pasal[34]\n'

In [ ]:
indices_1 = [3, 10, 17, 41, 43, 44, 45, 47]
indices_2 = [4, 13, 20, 24, 29, 34]

In [ ]:
for i in range(len(list_pasal)):
  if i in indices_1:
    temp = list_pasal[i].partition("  \n \nBAB")
    list_pasal[i] = temp[0]
  elif i in indices_2:
    temp = list_pasal[i].partition("  \n \nBagian Kedua")
    list_pasal[i] = temp[0]

In [ ]:
list_pasal[3]

'JKP sebagaimana dimaksud dalam Pasal 2 diselenggarakan \noleh BPJS  Ketenagakerjaan dan Pemerintah  Pusat .'

In [ ]:
list_pasal_modified = []
for i in range(2, len(list_pasal)):
  # remove multiple \n and whitespace
  temp = re.sub(' +', ' ', list_pasal[i].replace('\n', ' ')).strip()
  # modify the text so that (hopefully) the LLM can be instructed to easily identify the article number
  list_pasal_modified.append("[Berikut adalah isi Pasal " + str(i) + " pada " + "PP 37/2021" + "] " + temp)

In [ ]:
# List yang mengandung seluruh potongan peraturan
PP_37_2021 = pasal_1_modified + list_pasal_modified
len(PP_37_2021)

68

In [ ]:
PP_37_2021

['[Berikut adalah isi Pasal 1 angka 1 pada PP 37/2021] Jaminan Kehilangan Pekerjaan yang selanjutnya disingkat JKP adalah jamina n sosial yang diberikan kepada Pekerja/Buruh yang mengalami Pemutusan Hubungan Kerja berupa manfaat uang tunai , akses informasi pasar kerja, dan Pelatihan Kerja.',
 '[Berikut adalah isi Pasal 1 angka 2 pada PP 37/2021] Pekerja/Buruh adalah setiap orang yang bekerja dengan menerima Upah atau imbalan dalam bentuk lain.',
 '[Berikut adalah isi Pasal 1 angka 3 pada PP 37/2021] Pengusaha adalah: a. orang perseorangan, persekutuan, atau badan hukum yan g menjalankan suatu perusahaan milik sendiri; b. orang perseorangan, persekutuan, atau badan hukum yang secara berdiri sendiri menjalankan perusahaan bukan miliknya; c. orang perseorangan, persekutuan, atau badan hukum yang berada di Indonesia mewakili perusaha an sebagaimana dimaksud dalam huruf a dan huruf b yang berkedudukan di luar wilayah Indonesia.',
 '[Berikut adalah isi Pasal 1 angka 4 pada PP 37/2021] Upah 

## Data Loading

In [ ]:
# Menuliskan data ke dalam file .txt supaya dapat dipergunakan untuk keperluan lain
with open('list_of_data_PP_37_2021.txt', 'w') as outfile:
  outfile.writelines(["{}\n".format(item) for item in PP_37_2021])